README.md
koa-roles
=======

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![Gittip][gittip-image]][gittip-url]
[![David deps][david-image]][david-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/koa-roles.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-roles
[travis-image]: https://img.shields.io/travis/koajs/koa-roles.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/koa-roles
[coveralls-image]: https://img.shields.io/coveralls/koajs/koa-roles.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/koajs/koa-roles?branch=master
[gittip-image]: https://img.shields.io/gittip/fengmk2.svg?style=flat-square
[gittip-url]: https://www.gittip.com/fengmk2/
[david-image]: https://img.shields.io/david/koajs/koa-roles.svg?style=flat-square
[david-url]: https://david-dm.org/koajs/koa-roles
[download-image]: https://img.shields.io/npm/dm/koa-roles.svg?style=flat-square
[download-url]: https://npmjs.org/package/koa-roles

koa version of [connect-roles](https://github.com/ForbesLindesay/connect-roles)

## Install

```bash
$ npm install koa-roles
```

## Usage

```js
const Roles = require('koa-roles');
const Koa = require('koa');
const Router = require('koa-router');

const app = new Koa();
const router = new Router();
const user = new Roles({
  async failureHandler(ctx, action) {
    // optional function to customise code that runs when
    // user fails authorisation
    ctx.status = 403;
    var t = ctx.accepts('json', 'html');
    if (t === 'json') {
      ctx.body = {
        message: 'Access Denied - You don\'t have permission to: ' + action
      };
    } else if (t === 'html') {
      ctx.render('access-denied', {action: action});
    } else {
      ctx.body = 'Access Denied - You don\'t have permission to: ' + action;
    }
  }
});

app.use(user.middleware());
app.use(router.routes())
  .use(router.allowedMethods());
  
// anonymous users can only access the home page
// returning false stops any more rules from being
// considered
user.use(async (ctx, action) => {
  return ctx.user || action === 'access home page';
});

// moderator users can access private page, but
// they might not be the only ones so we don't return
// false if the user isn't a moderator
user.use('access private page', ctx => {
  if (ctx.user.role === 'moderator') {
    return true;
  }
})

//admin users can access all pages
user.use((ctx, action) => {
  if (ctx.user.role === 'admin') {
    return true;
  }
});

router.get('/', user.can('access home page'), async ctx => {
  await ctx.render('private');
});
router.get('/private', user.can('access private page'), async ctx => {
  await ctx.render('private');
});
router.get('/admin', user.can('access admin page'), async ctx => {
  await ctx.render('admin');
});

app.listen(3000);
```

## License

[MIT](LICENSE.txt)


History.md

2.0.0 / 2017-11-23
==================

**others**
  * [[`ddc662e`](http://github.com/koajs/koa-roles/commit/ddc662e1ceab46205962de395aa7d1fd1b77d146)] -  refactor: [BREAK] use async function and support koa@2 (#13) (Yiyu He <<dead_horse@qq.com>>)

1.0.3 / 2017-01-25
==================

  * fix: use yield instead of yield* (#12)

1.0.2 / 2015-12-10
==================

 * chore: add node 4 and 5 ci
 * fix: role with same action name will override now

1.0.1 / 2015-08-23
==================

 * chore: use eslint
 * chore: remove path-to-regexp dependency
 * doc: Add JSDoc format documentation.

1.0.0 / 2015-05-12
==================

 * upgrade path-to-regexp

0.1.0 / 2014-06-11
==================

 * update deps
 * fix typo
 * validthis: true
 * update readme

0.0.1 / 2014-03-27
==================

 * add logo
 * first commit


test/koa-roles.test.js
/**!
 * koa-roles - test/koa-roles.test.js
 *
 * Copyright(c) 2014 fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

'use strict';

/**
 * Module dependencies.
 */

require('should');
const request = require('supertest');
const koa = require('koa');
const sleep = require('co-sleep');
const Router = require('koa-router');
const pedding = require('pedding');
const Roles = require('../');

describe('koa-roles.test.js', function() {
  let app = new koa();
  const roles = new Roles();

  roles.use('every one', async function() {
    return true;
  });

  roles.use('user or admin', async ctx => {
    await sleep(1);
    return ctx.query.role === 'user' || ctx.query.role === 'admin';
  });

  roles.use('employee', ctx => {
    const role = ctx.query.role;
    return new Promise(function(resolve) {
      setTimeout(function() {
        resolve(role === 'employee');
      }, 1000);
    });
  });

  roles.use('update', ctx => {
    return ctx.query.role === 'user';
  });

  roles.use('user', ctx => {
    return ctx.query.role === 'user';
  });

  roles.use('friend', ctx => {
    return ctx.query.role === 'shaoshuai0102';
  });

  // override previous friend
  roles.use('friend', ctx => {
    return ctx.query.role === 'bar';
  });

  // default
  roles.use((ctx, action) => {
    if (ctx.query.role === action) {
      return true;
    }
  });

  // other default role check using generator function
  roles.use(async (ctx, action) => {
    await sleep(2);
    if (ctx.query.role2 === action) {
      return true;
    }
  });

  // using async function
  roles.use((ctx, action) => {
    const role = ctx.query.role3;
    return new Promise(function(resolve) {
      setTimeout(function() {
        resolve(role === action);
      }, 1000);
    });
  });

  app.use((ctx, next) => {
    if (ctx.query.role) {
      ctx.user = {};
      ctx.locals = {};
    }
    return next();
  });

  app.use(roles.middleware());

  const router = new Router();
  app.use(router.routes());

  router.get('/', roles.can('every one'), async ctx => {
    ctx.body = 'page for every one can visit';
  });

  router.get('/admin', roles.can('admin'), async ctx => {
    ctx.body = 'page only for admin can visit';
  });

  router.get('/admin/employee', roles.can('employee'), async ctx => {
    ctx.body = 'page for employee can visit';
  });

  router.get('/user', roles.is('user'), async ctx => {
    ctx.body = 'page only for user';
  });

  router.post('/profile/:id', roles.can('update'), async ctx => {
    ctx.body = 'page for user update';
  });

  router.get('/profile/:id', roles.can('user or admin'), async ctx => {
    ctx.body = 'page can visit by user or admin, current is ' + ctx.query.role;
  });

  router.get('/friend', roles.can('friend'), async ctx => {
    ctx.body = 'The best friend of foo is ' + ctx.query.role;
  });

  router.get('/any', async ctx => {
    const isadmin = await ctx.userCan('admin');
    if (!isadmin) {
      return ctx.throw(403);
    }
    ctx.body = 'hello admin';
  });

  before(function(done) {
    app = app.listen(0, done);
  });

  it('should throw Error when arguments bigger than 2', function() {
    (function() {
      roles.use(1, 2, 3);
    }).should.throw('use can have 1 or 2 arguments, not 3');
  });

  it('should throw Error when action start /', function() {
    (function() {
      roles.use('/foo', function() {});
    }).should.throw('action can\'t start with `/`');
  });

  it('should throw Error when action is not string', function() {
    (function() {
      roles.use(1, function() {});
    }).should.throw('Expected action to be of type string');
  });

  it('should throw Error when fn is not function', function() {
    (function() {
      roles.use('foo');
    }).should.throw('Expected fn to be of type function or generator function');
  });

  it('should get / 200 for every one', function(done) {
    request(app)
      .get('/')
      .expect('page for every one can visit')
      .expect(200, done);
  });

  it('should get / 200 for admin', function(done) {
    done = pedding(3, done);
    request(app)
      .get('/?role=admin')
      .expect('page for every one can visit')
      .expect(200, done);

    request(app)
      .get('/?role2=admin')
      .expect('page for every one can visit')
      .expect(200, done);

    request(app)
      .get('/?role3=admin')
      .expect('page for every one can visit')
      .expect(200, done);
  });

  it('should get / 200 for employee', function(done) {
    request(app)
      .get('/admin/employee?role=employee')
      .expect('page for employee can visit')
      .expect(200, done);
  });

  it('should get /admin 200 for admin', function(done) {
    done = pedding(4, done);
    request(app)
      .get('/admin?role=admin')
      .expect('page only for admin can visit')
      .expect(200, done);

    request(app)
      .get('/admin?role2=admin')
      .expect('page only for admin can visit')
      .expect(200, done);

    request(app)
      .get('/admin?role2=admin2')
      .set('accept', 'html')
      .expect('Access Denied - You don\'t have permission to: admin')
      .expect(403, done);

    request(app)
      .get('/admin?role2=admin2')
      .set('accept', 'application/json')
      .expect({ message: 'Access Denied - You don\'t have permission to: admin' })
      .expect(403, done);
  });

  it('should get /profile/:id 200 for admin or user', function(done) {
    done = pedding(3, done);
    request(app)
      .get('/profile/101?role=admin')
      .expect('page can visit by user or admin, current is admin')
      .expect(200, done);

    request(app)
      .get('/profile/1?role=user')
      .expect('page can visit by user or admin, current is user')
      .expect(200, done);

    request(app)
      .get('/profile/2?role2=admin2')
      .set('accept', 'application/json')
      .expect({ message: 'Access Denied - You don\'t have permission to: user or admin' })
      .expect(403, done);
  });

  it('should get /user 200 for user', function(done) {
    done = pedding(3, done);
    request(app)
      .get('/user?role=user')
      .expect('page only for user')
      .expect(200, done);

    request(app)
      .get('/user?role=admin')
      .expect(403, done);

    request(app)
      .get('/user')
      .expect(403, done);
  });

  it('should get /any 200 for admin', function(done) {
    done = pedding(2, done);
    request(app)
      .get('/any?role=admin')
      .expect('hello admin')
      .expect(200, done);

    request(app)
      .get('/any')
      .expect('Forbidden')
      .expect(403, done);
  });

  it('should post /profile 200 for user update', function(done) {
    done = pedding(3, done);
    request(app)
      .post('/profile/1?role=user')
      .expect('page for user update')
      .expect(200, done);

    request(app)
      .post('/profile/1?role=other')
      .expect({ message: 'Access Denied - You don\'t have permission to: update' })
      .expect(403, done);

    request(app)
      .post('/profile/1/123?role=other')
      .expect(404, done);
  });

  it('should get /friend 200 for bar', function(done) {
    request(app)
      .get('/friend?role=bar')
      .expect('The best friend of foo is bar')
      .expect(200, done);
  });

  it('should get /friend 403 for shaoshuai0102 because of role override', function(done) {
    request(app)
      .get('/friend?role=shaoshuai0102')
      .expect({ message: 'Access Denied - You don\'t have permission to: friend' })
      .expect(403, done);
  });
});


index.js
'use strict';

const is = require('is-type-of');

module.exports = KoaRoles;

/**
 * Role Middleware for Koa
 * @class KoaRoles
 * @example
 * ```js
 * var Roles = require('koa-roles');
 * var koa = require('koa');
 * var app = koa();
 *
 * var user = new Roles({
 *   failureHandler: function *(action) {
 *     // optional function to customise code that runs when
 *     // user fails authorisation
 *     this.status = 403;
 *     var t = this.accepts('json', 'html');
 *     if (t === 'json') {
 *       this.body = {
 *         message: 'Access Denied - You don\'t have permission to: ' + action
 *       };
 *     } else if (t === 'html') {
 *       this.render('access-denied', {action: action});
 *     } else {
 *       this.body = 'Access Denied - You don\'t have permission to: ' + action;
 *     }
 *   }
 * });
 *
 * app.use(user.middleware());
 *
 * // anonymous users can only access the home page
 * // returning false stops any more rules from being
 * // considered
 * user.use(function *(action) {
 *   return action === 'access home page';
 * });
 *
 * // moderator users can access private page, but
 * // they might not be the only ones so we don't return
 * // false if the user isn't a moderator
 * user.use('access private page', function (action) {
 *   if (this.user.role === 'moderator') {
 *     return true;
 *   }
 * });
 * app.get('/', user.can('access home page'), function *(next) {
 *   this.render('private');
 * });
 * app.get('/private', user.can('access private page'), function *(next) {
 *   this.render('private');
 * });
 * app.get('/admin', user.can('access admin page'), function *(next) {
 *   this.render('admin');
 * });
 *
 * app.listen(3000);
 * ```
 */
function KoaRoles(options) {
  options = options || {};
  this.functionList = [];
  this.failureHandler = options.failureHandler || defaultFailureHandler;
  this.userProperty = options.userProperty || 'user';
  this.actionMap = {};
}

/**
 * @method KoaRoles#use
 * @param {String} action - Name of role
 * @param {Function} fn
 */
KoaRoles.prototype.use = function() {
  if (arguments.length === 1) {
    // role.use(fn);
    this.use1(arguments[0]);
  } else if (arguments.length === 2) {
    // role.use(action, fn);
    this.use2(arguments[0], arguments[1]);
  } else {
    throw new Error('use can have 1 or 2 arguments, not ' + arguments.length);
  }
};

function assertFunction(fn) {
  if (!is.function(fn)) {
    throw new TypeError('Expected fn to be of type function or generator function');
  }
}

KoaRoles.prototype.use1 = function(fn) {
  assertFunction(fn);
  this.functionList.push(fn);
};

KoaRoles.prototype.use2 = function(action, fn) {
  if (typeof action !== 'string') {
    throw new TypeError('Expected action to be of type string');
  }
  if (action[0] === '/') {
    throw new TypeError('action can\'t start with `/`');
  }
  assertFunction(fn);

  const old = this.actionMap[action];
  // create or override
  this.actionMap[action] = fn;

  // action fn have already been used, skip
  if (old) {
    return;
  }

  const roles = this;
  this.use1((ctx, act) => {
    // get fn from actionMap
    const fn = roles.actionMap[action];
    if (act === action) {
      return fn(ctx, act);
    }
  });
};

/**
 * @method KoaRoles#can
 * @param {String} action
 */
KoaRoles.prototype.can = function(action) {
  const roles = this;
  return async (ctx, next) => {
    if (await roles.test(ctx, action)) {
      return next();
    }
    const r = roles.failureHandler(ctx, action);
    if (is.promise(r)) await r;
  };
};

/**
 * @see KoaRoles#can
 * @method KoaRoles#is
 */
KoaRoles.prototype.is = KoaRoles.prototype.can;

/**
 * @method KoaRoles#test
 * @param {Context} ctx
 * @param {String} action
 */
KoaRoles.prototype.test = async function(ctx, action) {
  for (let i = 0; i < this.functionList.length; i++) {
    const fn = this.functionList[i];
    let vote = fn(ctx, action);
    if (is.promise(vote)) vote = await vote;

    if (typeof vote === 'boolean') return vote;
  }
  return false;
};

/**
 * @method KoaRoles#middleware
 */
KoaRoles.prototype.middleware = function(options) {
  options = options || {};
  const userProperty = options.userProperty || this.userProperty;
  const roles = this;
  return (ctx, next) => {
    const roleCheck = tester(roles, ctx);
    if (ctx[userProperty]) {
      ctx[userProperty].is = ctx[userProperty].can = roleCheck;
      if (ctx.locals && !ctx.locals[userProperty]) {
        ctx.locals[userProperty] = ctx[userProperty];
      }
    }
    ctx.userIs = ctx.userCan = roleCheck;
    return next();
  };
};

function tester(roles, ctx) {
  return action => roles.test(ctx, action);
}

function defaultFailureHandler(ctx, action) {
  const message = `Access Denied - You don't have permission to: ${action}`;
  ctx.body = ctx.accepts('json', 'html') === 'json' ? { message } : message;
  ctx.status = 403;
}
